src/main/java/org/openlmis/buq/web/buq/BottomUpQuantificationController.java

```diff
@@ -247,6 +256,51 @@ public ApproveFacilityForecastingStats getApproveFacilityForecastingStats(
     return bottomUpQuantificationService.getApproveFacilityForecastingStats(programId);
   }
 
+  /**
+   * Rejects given bottom-up quantification.
+   *
+   * @param bottomUpQuantificationId UUID of BottomUpQuantification to authorize.
+   * @param rejectionDto DTO of Rejection
+   * @return rejected BottomUpQuantification.
+   */
+  @PostMapping("/{id}/reject")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public BottomUpQuantificationDto reject(@PathVariable("id") UUID bottomUpQuantificationId,
+      @RequestBody RejectionDto rejectionDto) {
+    BottomUpQuantification updatedBottomUpQuantification = bottomUpQuantificationService
+            .reject(bottomUpQuantificationId, rejectionDto);
+    return bottomUpQuantificationDtoBuilder.buildDto(updatedBottomUpQuantification);
+  }
+
+  /**
+   * Returns the latest rejection.
+   *
+   * @param bottomUpQuantificationId UUID of BottomUpQuantification to authorize.
+   * @return rejectionDto
+   */
+  @GetMapping("/{id}/mostRecentRejection")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public RejectionDto getMostRecentRejection(@PathVariable("id") UUID bottomUpQuantificationId) {
+    BottomUpQuantification bottomUpQuantification =
+            bottomUpQuantificationService.findBottomUpQuantification(bottomUpQuantificationId);
+    List<BottomUpQuantificationStatusChange> rejectedStatuses = bottomUpQuantification
```

Comment:
Hmm, why this whole method body (almost) isn't done as repository query/inside service?
I mean something like `getLastRejection(BottomUpQuantification):Optional<Rejection>` with some maybe helpers in repository like `getStatusChangesForStatusDescendingByDate(BottomUpQuantification, BottomUpQuantificationStatus).setMaxResult(1)` or maybe it all could even be written as db query.

Reply:
you are right, I forgot to move the logic to the service - something was not working here and I started to adding a bunch of code forgetting to move it.
Will move it to a service method.
The db query sounds fine, but unfortunately since I don't have much time, I will only move it to the service if you are fine with it.

Reply:
I'm fine with moving it only to service.

Reply:
moved